doc(aws-sts): add alternative to host OIDC configuration in a private bucket using CloudFront

mtulio opened the pull request:
Add steps to expose the OIDC endpoint in CloudFront with a private S3 Bucket.

Ref:
- https://issues.redhat.com/browse/SPLAT-617
- https://issues.redhat.com/browse/CCO-221

mtulio commented:
/cc @akhil-rane

mtulio commented on docs/sts.md:
````diff
@@ -45,19 +45,28 @@ The token is a projected ServiceAccount into the Pod, and is short lived for an
+
 8. Copy the manifests created in the step 7 and put them in the same location as install-config.yaml in the `manifests` directory
    ```
-   cp _output/manifests/* /path/to/dir/with/install-config.yaml/manifests/
````
changed as no `./openshift-install` command is specifying the `--dir` to something like  `/path/to/dir/with/install-config.yaml`

mtulio replied:
ACK, reverted!

mtulio replied:
@akhil-rane I believe to keep full consise with that sentence, we need also to change the entire document following the same idea. Look at the steps 4, 5 and 6. I reverted that change in my PR as it could be outscoped, but I believe to avoid confusion and keep in sync we could adapt those topics too.

akhil-rane replied:
yes, we can do that. I believe you want to have something like  `./openshift-install create install-config --dir=/path/to/dir/of /install-config`. If yes, do you want to make that change as a separate commit in this PR?

mtulio replied:
@akhil-rane yes, I believe we can do it in one separate PR to ship the current one faster. I will keep this thread open and update it when I have the other created.

mtulio replied:
PR opened for this topic https://github.com/openshift/cloud-credential-operator/pull/471

mtulio commented:
Tests:

```bash
$ curl https://${CLOUDFRONT_URI}/.well-known/openid-configuration
{
	"issuer": "https://d2k8zi1lcrqfal.cloudfront.net",
	"jwks_uri": "https://d2k8zi1lcrqfal.cloudfront.net/keys.json",
    "response_types_supported": [
        "id_token"
    ],
    "subject_types_supported": [
        "public"
    ],
    "id_token_signing_alg_values_supported": [
        "RS256"
    ],
    "claims_supported": [
        "aud",
        "exp",
        "sub",
        "iat",
        "iss",
        "sub"
    ]
}

$ echo ${OIDC_ARN}
arn:aws:iam::${ACCOUNT_ID}:oidc-provider/d2k8zi1lcrqfal.cloudfront.net

$ ./ccoctl aws create-iam-roles \
>     --name=${CLUSTER_NAME} \
>     --region=${CLUSTER_REGION}\
>     --credentials-requests-dir=${DIR_CCO}/credrequests \
>     --identity-provider-arn=${OIDC_ARN} \
>     --output-dir ${DIR_CCO}
2022/06/08 00:57:49 Role arn:aws:iam::${ACCOUNT_ID}:role/ccodoc-zlj79-openshift-machine-api-aws-cloud-credentials created
2022/06/08 00:57:49 Saved credentials configuration to: /home/mtulio/openshift-labs/aws-cco-oidc-ccoctl-doc/_output/manifests/openshift-machine-api-aws-cloud-credentials-credentials.yaml
2022/06/08 00:57:49 Updated Role policy for Role ccodoc-zlj79-openshift-machine-api-aws-cloud-credentials
(....)

$ ./openshift-install create cluster --log-level=debug
(...)
DEBUG Time elapsed per stage:                      
DEBUG            cluster: 5m57s                    
DEBUG          bootstrap: 44s                      
DEBUG Bootstrap Complete: 9m18s                    
DEBUG                API: 1m40s                    
DEBUG  Bootstrap Destroy: 33s                      
DEBUG  Cluster Operators: 5m17s                    
INFO Time elapsed: 22m25s 

$ oc get authentication cluster -o json \
>     | jq .spec.serviceAccountIssuer
"https://d2k8zi1lcrqfal.cloudfront.net"

$ oc get co  -o json \
>     | jq -r ".items[].status.conditions[] | select(.type==\"Available\").status" \
>     | sort |uniq -c

     32 True

$ oc get co -o json \
>     | jq -r ".items[].status.conditions[] | select(.type==\"Degraded\").status" \
>     | sort |uniq -c
     32 False

$ oc get secrets aws-cloud-credentials \
>     -n openshift-machine-api \
>     -o jsonpath='{.data.credentials}' \
>     | base64 -d
[default]
role_arn = arn:aws:iam::${ACCOUNT_ID}:role/ccodoc-zlj79-openshift-machine-api-aws-cloud-credentials
web_identity_token_file = /var/run/secrets/openshift/serviceaccount/token

# Get Token path from AWS credentials mounted to pod
$ TOKEN_PATH=$(oc get secrets aws-cloud-credentials \
    -n openshift-machine-api \
    -o jsonpath='{.data.credentials}' |\
    base64 -d |\
    grep ^web_identity_token_file |\
    awk '{print$3}')

# Get Controler's pod
$ CAPI_POD=$(oc get pods -n openshift-machine-api \
    -l api=clusterapi \
    -o jsonpath='{.items[*].metadata.name}')

# Extract tokens from pod
$ TOKEN=$(oc exec -n openshift-machine-api ${CAPI_POD} \
    -c machine-controller -- cat ${TOKEN_PATH})

$ echo $TOKEN | awk -F. '{ print $2 }' | base64 -d 2>/dev/null | jq .iss
"https://d2k8zi1lcrqfal.cloudfront.net"

$ IAM_ROLE=$(oc get secrets aws-cloud-credentials \
    -n openshift-machine-api \
    -o jsonpath='{.data.credentials}' |\
    base64 -d |\
    grep ^role_arn |\
    awk '{print$3}')
```
- Assume role
```bash
$ aws sts assume-role-with-web-identity \
    --role-arn "${IAM_ROLE}" \
    --role-session-name "my-session" \
    --web-identity-token "${TOKEN}"
```

```json
{
    "Credentials": {
        "AccessKeyId": "ASIAT[redacted]",
        "SecretAccessKey": "[redacted]",
        "SessionToken": "[redacted]",
        "Expiration": "2022-06-08T05:29:37Z"
    },
    "SubjectFromWebIdentityToken": "system:serviceaccount:openshift-machine-api:machine-api-controllers",
    "AssumedRoleUser": {
        "AssumedRoleId": "AROAT[redacted]:my-session",
        "Arn": "arn:aws:sts::[redacted]:assumed-role/ccodoc-zlj79-openshift-machine-api-aws-cloud-credentials/my-session"
    },
    "Provider": "arn:aws:iam::[redacted]:oidc-provider/d2k8zi1lcrqfal.cloudfront.net",
    "Audience": "openshift"
}
```

jianping-shu commented:
I had a test with the procedure today. Here are some result/question:
(1)aws s3api create-bucket  failed, and it succeeded with appending  "--create-bucket-configuration LocationConstraint=us-east-2"
jianpingshu@jshu-mac installer-dir % aws s3api create-bucket \
    --bucket ${OIDC_BUCKET_NAME} \
    --acl private

An error occurred (IllegalLocationConstraintException) when calling the CreateBucket operation: The unspecified location constraint is incompatible for the region specific endpoint this request was sent to.

jianpingshu@jshu-mac installer-dir % aws s3api create-bucket --region us-east-2 \
    --bucket ${OIDC_BUCKET_NAME} \
    --acl private --create-bucket-configuration LocationConstraint=us-east-2
{
    "Location": "http://jshu-cluster1-8jq99-oidc.s3.amazonaws.com/"
}
(2) If destroy the cluster, if need to clear any resource manually besides "ccoctl delete "?

(3) After copying the manifests and tls, 
[cloud-user@preserve-for-hive-test manifests]$ ls
cluster-authentication-02-config.yaml
openshift-cloud-credential-operator-cloud-credential-operator-iam-ro-creds-credentials.yaml
openshift-cloud-network-config-controller-cloud-credentials-credentials.yaml
openshift-cluster-api-capa-manager-bootstrap-credentials-credentials.yaml
openshift-cluster-csi-drivers-ebs-cloud-credentials-credentials.yaml
openshift-image-registry-installer-cloud-credentials-credentials.yaml
openshift-ingress-operator-cloud-credentials-credentials.yaml
openshift-machine-api-aws-cloud-credentials-credentials.yaml
[cloud-user@preserve-for-hive-test manifests]$ vi openshift-cluster-api-capa-manager-bootstrap-credentials-credentials.yaml

the OCP cluster install failed with the following logs:
ERROR Cluster operator etcd Degraded is True with UpgradeBackupController_Error: UpgradeBackupControllerDegraded: unable to retrieve cluster version, no completed update was found in cluster version status history: [{Partial 2022-06-10 13:03:26 +0000 UTC <nil> 4.11.0-0.nightly-2022-06-06-201913 registry.ci.openshift.org/ocp/release@sha256:49af2f8eeef5a24b2418aa6ba0be097a6e74bc747d35d403afe51ff1b173fa0b false }] 
ERROR Cluster operator ingress Degraded is True with IngressDegraded: The "default" ingress controller reports Degraded=True: DegradedConditions: One or more other status conditions indicate a degraded state: LoadBalancerReady=False (LoadBalancerPending: The LoadBalancer service is pending), CanaryChecksSucceeding=False (CanaryChecksRepetitiveFailures: Canary route checks for the default ingress controller are failing) 
ERROR Cluster operator monitoring Degraded is True with UpdatingPrometheusOperatorFailed: Failed to rollout the stack. Error: updating prometheus operator: reconciling Prometheus Operator Deployment failed: creating Deployment object failed: Internal error occurred: admission plugin "image.openshift.io/ImagePolicy" failed to complete mutation in 13s

The bootstrap logs
Jun 10 13:26:35 ip-10-0-5-124 bootkube.sh[2546]: [#2507] failed to create some manifests:
Jun 10 13:26:35 ip-10-0-5-124 bootkube.sh[2546]: "openshift-cluster-api-capa-manager-bootstrap-credentials-credentials.yaml": failed to create secrets.v1./capa-manager-bootstrap-credentials -n openshift-cluster-api: namespaces "openshift-cluster-api" not found
Jun 10 13:26:35 ip-10-0-5-124 cluster-bootstrap[7096]: "openshift-cluster-api-capa-manager-bootstrap-credentials-credentials.yaml": failed to create secrets.v1./capa-manager-bootstrap-credentials -n openshift-cluster-api: namespaces "openshift-cluster-api" not found
Jun 10 13:26:35 ip-10-0-5-124 cluster-bootstrap[7096]: Failed to create "openshift-cluster-api-capa-manager-bootstrap-credentials-credentials.yaml" secrets.v1./capa-manager-bootstrap-credentials -n openshift-cluster-api: namespaces "openshift-cluster-api" not found

jianping-shu commented:
Looks like install fail is due to https://bugzilla.redhat.com/show_bug.cgi?id=2035903, I'll have one more try next week. 
PTAL the first 2 items.

jianping-shu commented:
I tried to install the cluster again following the procedure, the cluster was installed but no worker node was created so finally some operators are degraded. 
[cloud-user@preserve-for-hive-test manifests]$ oc get clusterversion
NAME      VERSION   AVAILABLE   PROGRESSING   SINCE   STATUS
version             False       True          17h     Unable to apply 4.11.0-0.nightly-2022-06-06-201913: some cluster operators have not yet rolled out

I have some findings (aws s3api create-bucket was executed successfully on linux machine this time so my previous comment on it can be ignored):
1. BUG: there are 7 credrequests now and 1 is with TechPreviewNoUpgrade(0000_30_cluster-api_00_credentials-request.yaml), if I ran the following command, ccoctl won't bypass it without --enable-tech-preview
[cloud-user@preserve-for-hive-test _output]$ ccoctl aws create-iam-roles \
>     --name=${CLUSTER_NAME} \
>     --region=${CLUSTER_REGION}\
>     --credentials-requests-dir=${DIR_CCO}/credrequests \
>     --identity-provider-arn=${OIDC_ARN} \
>     --output-dir ${DIR_CCO}
2022/06/10 08:28:14 Role arn:aws:iam::301721915996:role/jshu-cluster1-8jq99-openshift-cluster-api-capa-manager-bootstrap created
2022/06/10 08:28:14 Saved credentials configuration to: /home/cloud-user/test_case/cco_219/_output/manifests/openshift-cluster-api-capa-manager-bootstrap-credentials-credentials.yaml
2022/06/10 08:28:14 Updated Role policy for Role jshu-cluster1-8jq99-openshift-cluster-api-capa-manager-bootstrap

2. For procedure, extract credreqs was done in step 2 so it needn't to be repeated in the step of create-iam-roles. The credentials-requests-dir parameter for ccoctl aws create-iam-roles shall be updated accordingly.

3. After destroy cluster, CloudFront Distribution still exists. Shall we provide the procedure to cleanup it and other resources(if any)?

jianping-shu commented:
I checked the install failures, seem my created tokens didn't work, any suggestion is appreciated!

[cloud-user@preserve-for-hive-test ~]$ curl https://${CLOUDFRONT_URI}/.well-known/openid-configuration
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>PermanentRedirect</Code><Message>The bucket you are attempting to access must be addressed using the specified endpoint. Please send all future requests to this endpoint.</Message><Endpoint>s3.amazonaws.com</Endpoint><Bucket>jshu-clus1-glp5p-oidc</Bucket><RequestId>5P460MD02JHB307S</RequestId><HostId>1o7iQfv5ndngH1JTEfgPgATFksHxs3iVWdEOCUHr5m+jXOHyRGw0zF5yfBW9eNUglM/9+rEg2NU=</HostId>

[cloud-user@preserve-for-hive-test run2]$ aws sts assume-role-with-web-identity \
>     --role-arn "${IAM_ROLE}" \
>     --role-session-name "my-session" \
>     --web-identity-token "${TOKEN}"

An error occurred (InvalidIdentityToken) when calling the AssumeRoleWithWebIdentity operation: Couldn't retrieve verification key from your identity provider,  please reference AssumeRoleWithWebIdentity documentation for requirements

akhil-rane reviewed (commented):
@mtulio I tested the steps mentioned in the docs and it works perfectly. Thanks for the detailed documentation. I have added some comments/suggestions. PTAL

akhil-rane commented:
> I tried to install the cluster again following the procedure, the cluster was installed but no worker node was created so finally some operators are degraded.

@jianping-shu I was able to install a cluster following the steps in this documentation. I think some configuration issue on your end is causing the installation to fail.

akhil-rane commented:
> I checked the install failures, seem my created tokens didn't work, any suggestion is appreciated!

tokens are working fine for me

mtulio commented on docs/sts-private-bucket.md:
```diff
@@ -0,0 +1,241 @@
+
+Steps to create the Origin Access Identity (OAI) to be used to access the bucket through CloudFront Distribution:
+
+- Create the OAI and set the variable `OAI_CLODUFRONT_ID`:
```
```suggestion
- Create the OAI and set the variable `OAI_CLOUDFRONT_ID`:
```

mtulio commented on docs/sts-oidc-bucket-policy.json.tpl:
```diff
@@ -0,0 +1,15 @@
+            "Sid": "1",
+            "Effect": "Allow",
+            "Principal": {
+                "AWS": "arn:aws:iam::cloudfront:user/CloudFront Origin Access Identity ${OAI_CLODUFRONT_ID}"
```
```suggestion
                "AWS": "arn:aws:iam::cloudfront:user/CloudFront Origin Access Identity ${OAI_CLOUDFRONT_ID}"
```

mtulio commented on docs/sts-oidc-cloudfront.json.tpl:
```diff
@@ -0,0 +1,91 @@
+                    "Quantity": 0
+                },
+                "S3OriginConfig": {
+                    "OriginAccessIdentity": "origin-access-identity/cloudfront/${OAI_CLODUFRONT_ID}"
```
```suggestion
                    "OriginAccessIdentity": "origin-access-identity/cloudfront/${OAI_CLOUDFRONT_ID}"
```

mtulio reviewed (commented):
s/OAI_CLODUFRONT_ID/OAI_CLOUDFRONT_ID/g

mtulio commented on docs/sts-oidc-cloudfront.json.tpl:
```diff
@@ -0,0 +1,91 @@
+{
+    "DistributionConfig": {
+        "CallerReference": "${CLUSTER_NAME}",
```
This field is used to identify the OAI. See the line of OAI creation: https://github.com/openshift/cloud-credential-operator/pull/468/files#diff-58642afcdbf995583bb30d55f53f8ff7fdca66e74928ac4ef86f991350d2ee35R47

mtulio commented on docs/sts-oidc-cloudfront.json.tpl:
```diff
@@ -0,0 +1,91 @@
+            "CloudFrontDefaultCertificate": true
+        }
+    },
+    "Tags": {
+        "Items": [
+            {
+                "Key": "Name",
+                "Value": "${CLUSTER_NAME}"
+            }
+        ]
+    }
```
No at this document (create flow). But it can be useful in two ways:
- Identify the CloudFront reference on the Console
- The destroy flow can use that tag to lookup (I believe we can add more standards tags here, wdyt?)

akhil-rane replied:
Makes sense. Lets keep these for now. We can add other standard tags with ccoctl process.

mtulio commented on docs/sts-private-bucket.md:
````diff
@@ -0,0 +1,241 @@
+
+- Create the private Bucket
+
+```bash
+aws s3api create-bucket \
+    --bucket ${OIDC_BUCKET_NAME} \
+    --acl private
+```
````
Ok, thanks to share it. I just added the region argument.

mtulio commented:
@akhil-rane thanks for your review.
My last update ✅ :
- Reviewed your comments
- Fixed the Location when creating the Bucket
- Tested on `us-east-2`
- Added the Destroy flow
- Tested the install && destroy flow in  `us-east-2` successfully

mtulio commented:
@jianping-shu : tested in the same region as yours. Please review the steps as described in the latest documentation.

```
./openshift-install create cluster --log-level=debug
(...)
DEBUG Time elapsed per stage:                      
DEBUG            cluster: 6m18s                    
DEBUG          bootstrap: 50s                      
DEBUG Bootstrap Complete: 8m46s                    
DEBUG                API: 1m25s                    
DEBUG  Bootstrap Destroy: 1m7s                     
DEBUG  Cluster Operators: 6m51s                    
INFO Time elapsed: 24m28s  

$ curl https://${CLOUDFRONT_URI}/.well-known/openid-configuration
{
	"issuer": "https://d1ede1lkq4zzno.cloudfront.net",
	"jwks_uri": "https://d1ede1lkq4zzno.cloudfront.net/keys.json",

(...)

$ oc --kubeconfig auth/kubeconfig get authentication cluster -o json | jq .spec.sericeAccountIssuer
"https://d1ede1lkq4zzno.cloudfront.net"

$ oc --kubeconfig auth/kubeconfig get infrastructure cluster -o json | jq .status.platformStatus
{
  "aws": {
    "region": "us-east-2"
  },
  "type": "AWS"
}

$ oc --kubeconfig auth/kubeconfig get clusterversion
NAME      VERSION   AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.10.16   True        False         68m     Cluster version is 4.10.16


$ oc --kubeconfig auth/kubeconfig get co  -o json | jq -r ".items[].status.conditions[] | select(.type==\"Available\").status" | sort |uniq -c
     32 True

$ oc --kubeconfig auth/kubeconfig get co  -o json | jq -r ".items[].status.conditions[] | select(.type==\"Degraded\").status" | sort |uniq -c
     32 False

$ aws sts assume-role-with-web-identity \
>     --role-arn "${IAM_ROLE}" \
>     --role-session-name "my-session" \
>     --web-identity-token "${TOKEN}"
{
    "Credentials": {
        "AccessKeyId": "ASIAT[[redacted]]",
        "SecretAccessKey": "[redacted]",
        "SessionToken": "[redacted]",
        "Expiration": "2022-06-17T18:30:26Z"
    },
    "SubjectFromWebIdentityToken": "system:serviceaccount:openshift-machine-api:machine-api-controllers",
    "AssumedRoleUser": {
        "AssumedRoleId": "AROAT[redacted]:my-session",
        "Arn": "arn:aws:sts::[redacted]:assumed-role/ccodoc-use2-457h2-openshift-machine-api-aws-cloud-credentials/my-session"
    },
    "Provider": "arn:aws:iam::[redacted]:oidc-provider/d1ede1lkq4zzno.cloudfront.net",
    "Audience": "openshift"
}

$ ./openshift-install destroy cluster --log-level=debug

(++> Delete steps as described on the doc)
```

jianping-shu commented:
@akhil-rane @mtulio Thanks for the inputs! 
I tested again following the updated procedure, it worked fine now (seems aws api server is not very stable, some aws commands will succeed after multiple tries). The ccoctl also worked correctly to ignoring the "tech-preview" CR. The updated procedure looks good to me. There's section of deleting the resource after cluster uninstall, great!
I guess my previous failure is due to some env variables not exported correctly.

akhil-rane commented on docs/sts-private-bucket.md:
```diff
@@ -240,3 +255,113 @@ echo ${OIDC_ARN}
+
+## Steps to Delete<a name="steps-delete"></a>
+
+Those steps should be followed after you've removed the resources created by ccoctl described in [the delete section](./ccoctl.md#aws-delete).
```
```suggestion
These steps should be followed after you've removed the resources created by ccoctl described in [the delete section](./ccoctl.md#aws-delete).
```

akhil-rane commented on docs/sts-private-bucket.md:
```diff
@@ -240,3 +255,113 @@ echo ${OIDC_ARN}
+
+If the `ccoctl delete` command failed due to non-empty Bucket (`BucketNotEmpty`), you should follow those steps to complete the Bucket removal.
+
+- Clean the Bucket objects
```
```suggestion
- Remove the Bucket objects
```

akhil-rane commented on docs/sts-private-bucket.md:
````diff
@@ -240,3 +255,113 @@ echo ${OIDC_ARN}
+- Get the new ETag (it's also returned on the last command).
+
+> The last command updates the `ETAG` variable referencing to a new configuration (Disabled distribution). If you didn't note it you need to re-run the `get-distribution-config` command, as desribed above.
+
+```bash
+aws cloudfront delete-distribution \
+    --id ${DISTRIBUTION_ID} \
+    --if-match ${ETAG}
+```
````
```suggestion
- Delete the CloudFront Distribution

```bash
aws cloudfront delete-distribution \
    --id ${DISTRIBUTION_ID} \
    --if-match ${ETAG}
```
```

mtulio replied:
@akhil-rane the suggestion above breaks the code block for markdown (closing the snippet)

akhil-rane commented:
@mtulio few more review comments but overall this looks good!

mtulio commented:
@akhil-rane review complete. Could you please evaluate it?

akhil-rane commented:
@mtulio this looks good to merge. Can we squash the commits into one or maybe two?

mtulio commented:
@akhil-rane done!

akhil-rane commented:
/lgtm

akhil-rane commented:
This is just documentation. No change to any functionality.

/label docs-approved
/label qe-approved
/label px-approved
